### DISCUSSION TITLE
Automatically return the 404 response when the specified file does not exist

### DISCUSSION THREAD

**Discussion by louiealmeda** [mockoon/mockoon#468] [Ideas] [answered]
We have setup our mocks in our directories like this:
```
Users
- GET.json
- POST.json
- Xq138f
 - GET.JSON
- Pm5eR
 - GET.JSON
```
And have a rule matcher set for that in mockoon
`GET: users/:userId`

File: `mocks/users/{{urlParam 'userId'}}/GET.json`

----
Now here's the issue:  

`GET /users/Xq138f` -> returns 200 with the GET.JSON file as body  (all good)
`GET /users/doesnotexist` -> returns 200 with `Error while serving the file content: ENOENT: no such file or directory, open 'mocks/users/doesnotexist/GET.json'` (could have been 404?)

Would it be possible to have a setting (or should this be default?), where mockoon responds with a 404 (with the template we set as 404) when the file does not exist?

Thank you.

**Comment by 255kb** [comment, marked as answer]
This is definitely on purpose. The idea of Mockoon is to be highly customizable while still easy to use. And I tried to avoid automated or hidden behaviors like returning a 404 when the file is not found and another status code was defined. That's why the app is quite tolerant regarding file presence, or JSON validity. 
So, I would say this default behavior shouldn't be changed but I get that it could be a little bit confusing in the first place. However, I am open to a new setting enabling automated 404. 
Another way you could achieve this would be to use multiple responses and rules, with one response for each "user" and a default one if the URL param doesn't match. Something like this: 
![image](https://user-images.githubusercontent.com/7489814/116880843-c8a42c00-ac22-11eb-8e52-45624bcf0f2e.png)
And you could keep your dynamic file path: `mocks/users/{{urlParam 'userId'}}/GET.json`

**Reply by louiealmeda** [reply]
Ahhh I agree, and it would be a breaking change if it were to be the default behavior...   

I've also tried using the rules to handle 404s before, but it's quite counter intuitive for that usecase as it becomes a whitelist. And the consumer now needs to be aware which requests patterns would respond a 404.

> The idea of Mockoon is to be highly customizable while still easy to use  

I definitely agree on this and it is really evident on the state of mockoon today. 🙂 I believe having the option to respond a specific status when the file is missing would further simplify a lot of setup and will enable mocking of 404s in a more expected way.

**Reply by 255kb** [reply]
yes, let's try to avoid breaking changes 😃 
But this could definitely be a setting at the route response level: 
![image](https://user-images.githubusercontent.com/7489814/117175051-b36c0080-adce-11eb-9435-2f2d3770c26a.png)
Feel free to open a new issue!

**Reply by louiealmeda** [reply]
Created #473 @255kb, not sure though if that's the exact approach you have in mind. Thanks 🙂

**Reply by 255kb** [reply]
Thanks! Let's discuss it there from now on then.
